text/0000-layout-packed-aligned.md

```diff
@@ -0,0 +1,182 @@
+the alignment of the struct fields are decreased to be M. Then, the base alignment of the type is
+increased to be N.
+
+When a `#[repr(packed(M))]` struct transitively contains a field with `#[repr(align(N))]` type, depending on the
+target triplet, either:
+- The field is first `pad_to_align`. Then, the field is added to the struct with alignment decreased to M. The packing requirement overrides the alignment requirement. (GCC, `#[repr(Rust)]`, `#[repr(C)]` on gnu targets, `#[repr(system)]` on non-windows targets), or
+- The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Historically the meaning of `#[repr(C)]` has been somewhat ambiguous. When someone puts `#[repr(C)]` on their struct, their intention could be one of three things:
+1. Having a target-independent and stable representation of the data structure for storage or transmission.
+2. FFI with C and C++ libraries compiled for the same target.
+3. Interoperability with operating system APIs.
+
+Today, `#[repr(C)]` is being used for all 3 scenarios because the user cannot create a `#[repr(C)]` struct with ambiguous layout between targets. However, this also means
+that there exists some C layouts that cannot be specified using `#[repr(C)]`.
+
+This RFC addresses use case 2 with `#[repr(C)]` and use case 3 with `#[repr(system)]`. For use case 1, people will have to seek alternative solutions such as `crABI` or
+protobuf. However, it could be a footgun if people continue to use `#[repr(C)]` for use case 1.
+
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This RFC clarifies that:
+- `repr(C)` must interoperate with the C compiler for the target.
+- `repr(system)` must interoperate with the operating system APIs for the target.
+- Similiar to Clang, `repr(C)` does not guarantee consistent layout between targets.
+
+Alternatively, we can also create syntax that allows the user to specify exactly which semantic to use when packed structs transitively contains aligned fields.
+For example, a new attribute: #[repr(align_override_packed(N))] that can be used when the behavior of the child overriding the parent alignment is desired.
+
+#[repr(align(N))] #[repr(packed)] can be used together to get the opposite behavior, parent/outer alignment wins.
+
+Explicitly specifying the pack/align semantic has the drawback of complicating FFI. For example, you might need two different definition files depending on the target.
+
+Therefore, a stable layout across compilation target should be relegated as future work.
+
+
+
+
+# Prior art
+[prior-art]: #prior-art
+
+Clang matches the Windows ABI for `x86_64-pc-windows-msvc` and matches the GCC ABI for `x86_64-pc-windows-gnu`.
+
+MinGW always uses the GCC ABI.
```

Comment:
Is there prior art for a compiler that can lay out types both using the Windows ABI and the GCC ABI for code within a single target? If yes, how are they distinguishing the two? If no, why does Rust need this ability?

Reply:
gcc apparently supports that by using the [`ms_struct` attribute](https://gcc.gnu.org/onlinedocs/gcc/x86-Type-Attributes.html#index-ms_005fstruct-type-attribute_002c-x86)

Reply:
So that would correspond tom in Rust
- have `repr(C)` on win-gnu targets match non-win targets
- have a separate window-only `repr(MS)` to ask for the msvc layout

Reply:
For a more full parallel to `extern "ABI"`, we should also support `repr(GCC)`. Then `repr(C)` is a sort of alias to `repr(GCC)` or `repr(MS)` chosen by the target, like `extern "C"` is an alias (strongly newtyped) to `"sysv64"`/`"win64"` (etc).

Reply:
clang [appears to also implement `[[gnu::ms_struct]]`](https://clang.llvm.org/docs/AttributeReference.html#ms-struct) has [an open PR implementing `[[gnu::gcc_struct]]`](https://github.com/llvm/llvm-project/pull/71148)